Async migrations UI refresh

tiina303 opened the pull request:
## Changes

Before
<img width="964" alt="Screen Shot 2022-01-24 at 14 50 04" src="https://user-images.githubusercontent.com/890921/150795004-0d7c3323-0cbb-429c-8909-f04e26d316cb.png">

After
<img width="1110" alt="Screen Shot 2022-01-25 at 00 44 28" src="https://user-images.githubusercontent.com/890921/150884383-3fde2700-2816-4124-a924-013db73f52fd.png">

<img width="175" alt="Screen Shot 2022-01-24 at 14 53 24" src="https://user-images.githubusercontent.com/890921/150795359-93863bd8-04a2-46dc-9664-dd086502e347.png">
<img width="191" alt="Screen Shot 2022-01-24 at 14 53 31" src="https://user-images.githubusercontent.com/890921/150795375-a6f291ca-b952-4a2b-b497-0bcc92e3fa43.png">

## How did you test this code?

existing CI + played with the UI

tiina303 committed:
Async migrations to auto resume

tiina303 commented on posthog/async_migrations/setup.py:
```diff
@@ -91,8 +91,8 @@ def kickstart_migration_if_possible(migration_name: str, applied_migrations: set
 
-    # start running 30 minutes from now
-    run_next_migration(migration_name)
+    # start running new migrations 30 minutes from now or continue existing ones immediately
```
I'm not sure we have a 30min delay actually, what does https://github.com/PostHog/posthog/blob/master/posthog/async_migrations/utils.py#L58 that delay do?

fuziontech replied:
I don't see it actually delaying for 30 minutes? `delay` is just a way to [trigger a celery task](https://docs.celeryproject.org/en/stable/userguide/calling.html#example)

Am I missing something about the 30 minutes delay?

tiina303 replied:
I think it was there initially and then got removed in a later PR. We should potentially re-visit the idea anyway. So the comment is wrong.

tiina303 commented:
NVM we already have a celery periodic task for this

tiina303 commented on frontend/src/scenes/instance/AsyncMigrations/index.tsx:
```diff
@@ -48,10 +49,16 @@ export function AsyncMigrations(): JSX.Element {
+                            <div>
+                                <PlayCircleOutlined
+                                    className="migration-btn success"
+                                    onClick={() => resumeMigration(asyncMigration.id)}
```
there's a problem here: people could click this multiple times, could be fixed in a follow-up. Potentially something similar to https://github.com/PostHog/posthog/blob/master/posthog/tasks/async_migrations.py#L60 checking the active task ids in `is_migration_resumable` function.

yakkomajuri replied:
let's not let the migration resume if it is in a running state. probably should go in this pr

tiina303 replied:
I started reworking this. I think we should: 
1. by default not roll back. 
2. In the UI have these options: running state -> stop ; errored state -> rollback & resume;

fuziontech commented on posthog/api/async_migration.py:
```diff
@@ -78,6 +82,15 @@ def trigger(self, request, **kwargs):
+    @action(methods=["POST"], detail=True)
+    def resume(self, request, **kwargs):
+        migration_instance = self.get_object()
+        resumable, error = is_migration_resumable(migration_instance)
+        if error:
+            return response.Response({"success": False, "error": error,}, status=400,)
```
I love this kind of error handling <3
![image](https://user-images.githubusercontent.com/391319/148125756-5a5eabeb-d81e-4a35-a233-8c20d7bb89e7.png)

Should this be a `400` or should it be a `500`? Feels like a `500` to me.

tiina303 replied:
```suggestion
        resumable, error = is_migration_resumable(migration_instance)
        if error:
            return response.Response({"success": False, "error": error,}, status=500,)
```

fuziontech reviewed (commented):
Only have questions about `400` vs `500` with regard to the return http status if there is an error getting the migration. Also a question about the 30 minute delay - I don't see it delayed anywhere in the celery task?

tiina303 commented on posthog/async_migrations/setup.py:
```diff
@@ -91,8 +91,8 @@ def kickstart_migration_if_possible(migration_name: str, applied_migrations: set
 
-    # start running 30 minutes from now
-    run_next_migration(migration_name)
+    # start running new migrations 30 minutes from now or continue existing ones immediately
```
```suggestion
    # start running new migrations or continue the existing one
```

tiina303 commented:
> Only have questions about 400 vs 500 with regard to the return http status if there is an error getting the migration. Also a question about the 30 minute delay - I don't see it delayed anywhere in the celery task?

updated to 500. Updated the comment. I think the delay was there initially and then got removed in a later PR. We should potentially re-visit the idea anyway.

yakkomajuri reviewed (commented):
I'm not particularly fond of the UI here, but happy to move fast

tiina303 commented:
There's an open issue for product design for now we can just land this so we'll have all the options available from the UI

yakkomajuri reviewed (commented):
The buttons look very messy imo. I'm happy to merge in favor of moving fast, but is there something else we can do quickly? Maybe add some spacing and ensure they stay side-by-side? 

Could also have a dropdown with the actions and a trigger button?

Also let's now name this column. "Actions" maybe? We can also probably drop "Celery task id" to free up space? That's very much for using in an automated way as opposed to e.g. "Query id" which would be more useful for manual debugging.

tiina303 committed:
partial check-in

tiina303 commented:
WIP  currently looks like this

<img width="979" alt="Screen Shot 2022-01-21 at 01 49 48" src="https://user-images.githubusercontent.com/890921/150445848-00b40705-036c-469b-a7cc-b6b82cce1d27.png">

tiina303 commented on frontend/src/scenes/instance/AsyncMigrations/AsyncMigrationDetails.tsx:
```diff
@@ -0,0 +1,51 @@
+            dataIndex: 'description',
+        },
+    ]
+    // Problem: the state is global and pressing the button below changes errors for all open pannels probably because
+    // the value asyncMigrationErrors is global and gets changed
+    // couldn't figure out how to load this on the opening of the details
```
how should I do this?

Twixes replied:
Currently the reducer `asyncMigrationErrors` gets its value from the result of `loadAsyncMigrationErrors`. And `loadAsyncMigrationErrors` returns data related to a specific single migration (`migrationId`), of type `AsyncMigrationError[]`.
One way this can be fixed is making `asyncMigrationErrors` a mapping based on migration IDs (`Record<AsyncMigration['id'], AsyncMigrationError[]>`). This way in `AsyncMigrationDetails` instead of `dataSource={asyncMigrationErrors}` you'd do `dataSource={asyncMigrationErrors[asyncMigration.id]}`.

Twixes replied:
1. You can call the loading action inside a `useEffect()` when the details component renders (which equals the row having been expanded).
2. You can look at `dashboardsModel`'s `rawDashboards` reducer for inspiration, it's the same idea of memoizing with an ID-to-data object.

mariusandra replied:
I suggested earlier that Tiina can also add `onRowExpanded` to the `expandable` object, and call that (if present) from the TableRow's expansion button's `onClick` handler.

tiina303 replied:
which of those two is better?

tiina303 replied:
went with `onRowExpanded`

tiina303 commented on frontend/src/scenes/instance/AsyncMigrations/index.tsx:
```diff
@@ -200,11 +238,17 @@ export function AsyncMigrations(): JSX.Element {
                                 loading={asyncMigrationsLoading}
                                 columns={columns}
                                 dataSource={asyncMigrations}
+                                expandable={{
+                                    expandedRowRender: function renderExpand(row) {
+                                        return row && <AsyncMigrationDetails asyncMigration={row} />
+                                    },
+                                    rowExpandable: ({ error_cnt }) => error_cnt > 0,
+                                }}
```
why does the rendering look weird, when it shouldn't be expandable?

Twixes replied:
Sorry, this is a bug I introduced! Fixed now on this branch

mariusandra replied:
The [coding convention dos & donts here](https://posthog.com/docs/contribute/coding-conventions#do-s--dont-s) suggest not using "index.tsx" as a filename.

Twixes committed:
Fix `rowExpandable` checks

mariusandra commented on frontend/src/scenes/instance/AsyncMigrations/asyncMigrationsLogic.ts:
```diff
@@ -81,6 +93,15 @@ export const asyncMigrationsLogic = kea<asyncMigrationsLogicType<AsyncMigration,
+            {
+                loadAsyncMigrationErrors: async (migrationId): Promise<AsyncMigrationError[]> => {
+                    console.log(`trying to fetch the errors for ${migrationId}`)
+                    return (await api.get(`api/async_migrations/${migrationId}/errors`)).results
```
loaders: {
    loadAsyncMigrationErrors: async (migrationId): Promise<AsyncMigrationError[]> => {
        const errorsForMigration = await api.get(`api/async_migrations/${migrationId}/errors`)).results
        return { ...values.asyncMigrationErrors, [migraitonId]: errorsForMigration }
    }
},

reducers:  {
    asyncMigrationErrorLoading: [
        {} as Record<number, boolean>, 
        {
            loadAsyncMigrationErrors: (state, migrationId) => {
                return { ...state, [migrationId]: true }
            },
            loadAsyncMigrationErrorsSuccess: (state, payload) => {
                const migrationId = ....?
                return { ...state, [migrationId]: false }
            }
            loadAsyncMigrationErrorsFailure: (state, payload) => {
                return { ...state, [migrationId]: false }
            }
        }
    ]
}

asyncMigrationErrors[id].length > 0 ? <Errors /> : "no errors"

tiina303 committed:
loading errors when clicking Refresh

tiina303 committed:
rename index files

tiina303 committed:
Load errors on row expansion

tiina303 commented on frontend/src/lib/components/LemonTable/TableRow.tsx:
```diff
@@ -32,10 +32,13 @@ function TableRowRaw<T extends Record<string, any>>({
-                                onClick={() => setIsRowExpanded((state) => !state)}
+                                onClick={() => {
+                                    setIsRowExpanded((state) => !state)
+                                    !isRowExpanded && expandable.onRowExpanded && expandable.onRowExpanded(record)
```
at this point `isRowExpanded` was false when I hit expand, but I'm not sure exactly when `setIsRowExpanded` state change happens

tiina303 committed:
nits

tiina303 commented on frontend/src/scenes/instance/AsyncMigrations/asyncMigrationsLogic.ts:
```diff
@@ -43,21 +49,42 @@ export interface AsyncMigration {
+                    return { ...state, [payload.payload]: false }
+                },
+                loadAsyncMigrationErrorsFailure: (state, payload) => {
+                    console.log(payload)
```
here's a problem - I don't have enough info here (i.e. I don't know what ID the loading function was called on)
<img width="717" alt="Screen Shot 2022-01-24 at 13 49 21" src="https://user-images.githubusercontent.com/890921/150786188-3f8a7184-6df2-40f6-b550-d07c6a4f8382.png">

Twixes replied:
Hm, maybe the API's error response (which we use as `errorObject` verbatim) should include a `lookup` property, which would be the lookup value that was used (in this case, migration ID).

tiina303 replied:
Fixed this by manually defining listeners & everything instead of using loaders

Twixes commented on frontend/src/lib/components/LemonTable/TableRow.tsx:
```diff
@@ -32,10 +32,13 @@ function TableRowRaw<T extends Record<string, any>>({
                                 type={isRowExpanded ? 'highlighted' : 'stealth'}
-                                onClick={() => setIsRowExpanded((state) => !state)}
+                                onClick={() => {
+                                    setIsRowExpanded((state) => !state)
+                                    !isRowExpanded && expandable.onRowExpanded && expandable.onRowExpanded(record)
```
It may be a bit clearer if you just run that as a side effect of `setIsRowExpanded`.
```suggestion
                                    setIsRowExpanded((state) => {
                                        if (!state) {
                                            expandable.onRowExpanded?.(record)
                                        }
                                        return !state
                                    })
```

tiina303 replied:
that feels hidden to me, I'd expect `setIsRowExpanded` to not do anything besides setting that state

Twixes commented on frontend/src/lib/components/LemonTable/types.ts:
```diff
@@ -67,4 +67,6 @@ export interface ExpandableConfig<T extends Record<string, any>> {
      */
     rowExpandable?: (record: T) => boolean | number
+    /** Called when row is expanded */
+    onRowExpanded?: (record: T) => null
```
A nit, but I suggest instead naming `onRowExpand`, as that'd be more inline with the convention arounds events (e.g. `onClick`, `onLoad`, `onChange`).

Twixes commented on frontend/src/scenes/instance/AsyncMigrations/AsyncMigrations.tsx:
```diff
@@ -18,70 +19,51 @@ import {
             dataIndex: 'name',
         },
+        {
+            title: 'Error cnt',
```
Does this mean "Error count"?

Twixes commented on frontend/src/scenes/instance/AsyncMigrations/AsyncMigrations.tsx:
```diff
@@ -18,70 +19,51 @@ import {
+        },
         {
             title: 'Description',
-            render: function RenderError(asyncMigration: AsyncMigration): JSX.Element {
+            render: function Render(_, asyncMigration: AsyncMigration): JSX.Element {
```
This column is very clunky, particularly because I have to click the tiny ellipsis button on each description just to read it to the end. Could we use a similar pattern as in Insights/Dashboards/Feature flags here, where the description is shown in the name column (class `row-description`) and not truncated?

tiina303 replied:
<img width="528" alt="Screen Shot 2022-01-24 at 15 47 44" src="https://user-images.githubusercontent.com/890921/150805158-c93bc468-2b6e-403a-ac40-791d0e657578.png">

Updated description based on the product design (https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=6625%3A32152). How I can make the length be tied to column length?

tiina303 replied:
not really I think the descriptions could be really long especially in the future & I think we should add more stuff there too

tiina303 replied:
also this doesn't need to be perfect yet :) I can continue to make it look nicer for next release

tiina303 replied:
ok this is what I got and it's probably fine for this PR we can improve it in future releases:
<img width="473" alt="Screen Shot 2022-01-24 at 18 22 50" src="https://user-images.githubusercontent.com/890921/150833379-a223697b-16bf-4bd8-8772-b1fca277b99d.png">

tiina303 committed:
don't reuse name ... typ errors

Twixes commented on posthog/api/async_migration.py:
```diff
@@ -9,11 +9,28 @@
+    error_cnt = serializers.SerializerMethodField()
+
+    def get_error_cnt(self, async_migration: AsyncMigration):
+        return len(AsyncMigrationError.objects.filter(async_migration=async_migration))
```
`<SELECT COUNT(*)>` instead of `len(<SELECT *>)`:
```suggestion
        return AsyncMigrationError.objects.filter(async_migration=async_migration).count()
```

tiina303 committed:
buttons refresh

tiina303 committed:
review nits

yakkomajuri commented:
@Twixes is already reviewing this PR and he's more qualified when it comes to our frontend than me, so I'll skip this review.

tiina303 committed:
description

tiina303 committed:
fix type error

tiina303 committed:
lemontag for status

tiina303 commented on frontend/src/scenes/instance/AsyncMigrations/AsyncMigrations.tsx:
```diff
@@ -0,0 +1,306 @@
+            render: function Render(_, asyncMigration: AsyncMigration): JSX.Element {
+                const status = asyncMigration.status
+                let type = 'default'
+                if (status === AsyncMigrationStatus.Errored) {type = 'danger'}
+                if (status === AsyncMigrationStatus.CompletedSuccessfully) {type = 'success'}
+                if (status === AsyncMigrationStatus.Running) {type = 'warning'}
```
these don't seem to work, but I'm not sure why?
<img width="250" alt="Screen Shot 2022-01-24 at 16 32 05" src="https://user-images.githubusercontent.com/890921/150813599-9ac6a1e0-9988-4176-9346-541de8e21b67.png">

tiina303 replied:
because of our css file that had defined .danger etc to use those colors, i.e. we were trying to write on red with red

tiina303 committed:
Description and more button rendering

tiina303 committed:
fix loading failure

tiina303 committed:
Fix status style

tiina303 committed:
forgotten include

tiina303 committed:
nits

mariusandra reviewed (commented):
The logic looks solid to me! Great work! 👍 

I could still find some UX things to comment on:
1. The inner table could definitely use some more whitespace around it, same for the refresh button:
![image](https://user-images.githubusercontent.com/53387/150965676-97a02ebd-9d28-4961-a443-8c81971d997e.png)

2. Not sure if the blue "[...]" is in the right place nor if the modal is the right way to show the data, and I didn't check all the states, but I'll defer to people more knowledgeable here.
![image](https://user-images.githubusercontent.com/53387/150966572-6f0fa977-34bb-4c03-a130-fa3a891f07fd.png)

tiina303 commented:
> Not sure if the blue "[...]" is in the right place nor if the modal is the right way to show the data, and I didn't check all the states, but I'll defer to people more knowledgeable here.

I think you might have been looking at the before pictures, check the last 3 pics.

mariusandra commented:
> > Not sure if the blue "[...]" is in the right place nor if the modal is the right way to show the data, and I didn't check all the states, but I'll defer to people more knowledgeable here.
> 
> I think you might have been looking at the before pictures, check the last 3 pics.

Awkward. You're right. Somehow I was running something older locally.

Twixes committed:
Merge branch 'master' into async-migrations-auto-resume

Twixes committed:
Use Material icons instead of AntD

Twixes committed:
Simplify async migration name/description

Twixes committed:
Center spinner in button

Twixes committed:
Use turbo mode

Twixes committed:
Fix errors table style

Twixes reviewed (approved):
Generally looks good. I also made a few stylistic improvements so that things are laid out well:

### Before
![Before](https://user-images.githubusercontent.com/4550621/150991871-d673a068-9a42-48f7-93e9-fe96d661e563.png)

### After
![After](https://user-images.githubusercontent.com/4550621/150991867-a53f272a-1f3b-4424-a002-f5f613b1e196.png)
